src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
```

Comment:
> The stdio module helpers should be there with the stdio.obj file, along with stdio's moduleinfo. 

Yes, it's still correct. When we compile stdio.d, stdio.obj will contain both non-COMDAT moduleinfo and three helper functions. 

> There should be no reason for any other module to generate either.

No, it's necessary to fix issue 846. When a user instantiate writeln!(), it will implicitly use the helper `stdio.__array` for array bounds check. But, if he don't link stdio.obj, it had got the linker error "undefined symbol stdio.__array". In #3552, I've applied those changes to fix 846:

- All module helper functions had changed to not use ModuleInfo. It was just for getting module file name, so I changed to use bare module name string as the alternative.
- After that, each helper functions has changed to not dependent to the corresponding ModuleInfo. When they're required for a non-root template instance, they will be also generated and placed in COMDAT section. Finally, the use of them won't cause link failures, even if the non-COMDAT versions are not linked.

If you want to know the details, please also check the change in #3552. Honestly this PR is a supplemental change of that. Honestly, when I wrote that PR, I didn't think well about the library generation. So the COMDAT helpers had placed in incorrect places. This PR will fix that issue.